hare: embed paths for cross-compilation toolchains

onemoresuza opened the pull request:
## Description of changes

Embed the needed paths on the `Hare` binary, so that it can cross-compile.

<!--
For package updates please link to a changelog or describe changes, this helps your fellow maintainers discover breaking updates.
For new packages please briefly describe the package or provide a link to its homepage.
-->

## Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [X] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [X] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [X] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [X] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [24.05 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) (or backporting [23.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md) and [23.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [X] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!

List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://nixos.org/manual/nixpkgs/unstable/#chap-reviewing-contributions
-->

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

onemoresuza committed:
hare: embedd paths for cross-compilation toolchains

Also strip the `HAREPATH` make variable embedded in the binary from
empty `$(SRCDIR)/hare/third-party`, since since nix does not follows the
FHS and relies on the `HAREPATH` environment variable to find third
party libraries.

onemoresuza commented:
Result of `nixpkgs-review` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>1 package marked as broken and skipped:</summary>
  <ul>
    <li>himitsu-firefox</li>
  </ul>
</details>
<details>
  <summary>9 packages built:</summary>
  <ul>
    <li>hare</li>
    <li>hare.man</li>
    <li>hareThirdParty.hare-compress</li>
    <li>hareThirdParty.hare-json</li>
    <li>haredo</li>
    <li>haredo.man</li>
    <li>himitsu</li>
    <li>treecat</li>
    <li>treecat.man</li>
  </ul>
</details>

onemoresuza commented:
@uninsane, I saw that you had to do some manual intervention for cross-compilation on your [bonsai PR](https://github.com/NixOS/nixpkgs/pull/233892#issuecomment-1827010523):

> [...] i took the liberty to add the necessary hare flags to support cross-compilation while i was at it though (i think the flags for that can be added to the hare setup hook, after we verify it works with the other hare packages).

Could you elaborate on that?

uninsane commented:
> Could you elaborate on that?

`nix build '.#pkgsCross.aarch64-multiplatform.bonsai'`

1. `HAREPATH` has to manually be set to the third-party hare packages (hare-json, hare-ev), else harec can't find them:
```
> hare build  -o bonsaid cmd/bonsaid/
> Error: /build/source/bonsai: encoding::json: Module not found
```

2. `HARE_TARGET_FLAGS` has to manually be set to "-a aarch64"`, else hare generates x86 assembly and qbe (correctly) fails on that:
```
Error: unknown mnemonic `pushq' -- `pushq %rbp'
Error: unknown mnemonic `movq' -- `movq %rsp,%rbp'
...
``` 

when i looked several months ago the nixpkgs hare setup hook is supposed to do (1) for you, but it only sees hare packages which are in `nativeBuildInputs`, not `buildInputs`. (2) probably belongs in the setup hook, but i wasn't sure if HARE_TARGET_FLAGS is supported universally by hare, or if that was something in the bonsai Makefile reading that variable.

i'll try out your solution here later when i have a bit more time.

onemoresuza commented:
I think this PR still not addresses the issues you encountered: What it does, as of now, is provide the toolchain for cross-compilation, *i. e.*, set the correct paths for make variables with the tools paths that are embedded in the `hare` binary.

I think both `1` and `2` may be addressed by the setup hook: the former just needs that hare's third party libraries are added to HAREPATH when they are not only `nativeBuildInputs`, but also on `buildInputs`; while the latter may be addressed by passing `a <arch>` to the `HAREFLAGS` environment variable.

uninsane commented on pkgs/by-name/ha/hare/setup-hook.sh:
```diff
@@ -1,9 +1,10 @@
 }
 
-addEnvHooks "$hostOffset" addHarepath
+addEnvHooks "${hostOffset:?}" addHarepath
+addEnvHooks "${targetOffset:?}" addHarepath
```
this seems suspicious to me: why should hare dependencies be drawn from more than just one offset?

looking at other setup-hooks which do dependency detection seems to confirm this (use `targetOffset`):
- pkgs/build-support/pkg-config-wrapper/setup-hook.sh
- pkgs/by-name/cm/cmake/setup-hook.sh

onemoresuza replied:
I've taken a look at your [bonsai PR](https://github.com/NixOS/nixpkgs/pull/233892/) and I think I can explain why you had to manually add the dependencies to `HAREPATH` and why I should revert this change.

`Hare` binaries, when written in pure `Hare` --- *i. e.*, no `hare build -l<c-lib>` --- are statically linked, so its third party libraries in `nixpkgs` should only be added to `nativeBuildInputs` --- build only dependencies --- and not to `buildInputs` --- runtime dependencies.

In the case of the aforementioned PR, moving `hare-json` and `hare-ev` from `buildInputs` to `nativeBuildInputs` will solve the problem.

`Hare` `Error: unknown mnemonic` errors are caused because it is using the `buildPlatform` architecture as its target platform. For cross-compilation, we must find a way to pass the `-a` flag to `Hare`.

uninsane commented:
can you clarify exactly what this PR is meant to enable? i tried `nix build '.#pkgsCross.aarch64-multiplatform.hare'`, and it still fails:
```
Error: unknown configure option --build=x86_64-unknown-linux-gnu
```

same with `harec`.

cross-compilation of hare packages from x86_64 (like bonsai, over [here](https://github.com/NixOS/nixpkgs/pull/233892)) already works without this, and the patch doesn't do anything to simplify that (you were right that `HARE_TARGET_FLAGS` was specific to the bonsai package).

is it specifically enabling cross compilation from non-x86_64 hosts?

onemoresuza commented on pkgs/by-name/ha/hare/cross-compilation-tests.nix:
```diff
@@ -0,0 +1,36 @@
+        *) ;;
+      esac
+    )"
+    ${fileBin} "$outbin" | grep "^$outbin:.*,[[:space:]]*$filearch,"
```
I see. Going to change the tests to use `stdenv.hostPlatform.emulator` and change the `Hare` code to check the architecture inside of it using [os::machine](https://docs.harelang.org/os#machine).

onemoresuza commented:
> can you clarify exactly what this PR is meant to enable? i tried `nix build '.#pkgsCross.aarch64-multiplatform.hare'`, and it still fails:
> 
> ```
> Error: unknown configure option --build=x86_64-unknown-linux-gnu
> ```
> 
> same with `harec`.

The error here is that `stdenv.mkDerivation` is passing an unknown option to `harec`'s config script `--build`. Actually, both `--build` and `--host` is being passed to it, since `[ "build" "host" ]` are the default values of `configurePlatforms` when cross-compiling. Setting it to `[ ]` fixes the problem.

But, by doing that, we have another problem: We are going to have a toolchain for the `buildPlatform` with the exception of `qbe`, which is built for the `hostPlatform`; so, when the configure script tries to find it by running `qbe -h`, it will not, since it cannot run the binary for such architecture.

> is it specifically enabling cross compilation from non-x86_64 hosts?

I think the PR name ended up being misleading: The cross-compilation being enabled is the one for the `Hare` binary itself outside of `nixpkgs`. So if one wishes to pull up a `nix-shell` with `hare` on it, there will be no need to configure the toolchain for other platforms.

To enable cross-compilation on `nixpkgs` for `Hare` packages, we must find a way to set the `-a` flag. Some packages have a `HAREFLAGS` make variable for it, but it is not enforceable. Perhaps this would be the moment to create a `buildHarePackage` derivation. It could:

1. Set `HARECACHE`, avoiding the need of packages to do it manually;
2. replace the `hare` command on the shell with something like this to allow cross-compilation to work:
    ```bash
    declare -r harecmd="${1}"
    declare -a newargs
    shift
    # Remove any mention of the `-a` flag, setting it latter according to
    # the `hostPlatform`.
    while [[ -n "${1}" ]]; do
      case "${1}" in
        -a?*) shift ;;
        -a) ;;
        *) newargs+=("${1}") ;;
      esac
      shift
    done
    [[ "${harecmd}" == "build" ]] && arch=@arch@ || arch=""
    hare "${harecmd}" ${arch:+-a${arch} "${newargs[@]}"
    ```

uninsane commented:
thanks for explaining! in that case, i've confirmed that this PR does fix workflows like this:
```console
$ cd ~/src/nixpkgs
$ git checkout master
$ nix build '.#hare' -o hare-master
$ git checkout hare-crosscomp
$ nix build '.#hare' -o hare-cross

$ git clone https://git.sr.ht/~autumnull/haredo
$ cd haredo
$ sed -i 's/hare build/hare build -a aarch64/' bootstrap.sh
$ PATH=$PATH:$(pwd)/../hare-master/bin ./bootstrap.sh
  2/97 tasks completed (2%)Error: Command not found: aarch64-as
$ PATH=$PATH:$(pwd)/../hare-cross/bin ./bootstrap.sh
$ file bin/haredo
bin/haredo: ELF 64-bit LSB executable, ARM aarch64, version 1 (SYSV), statically linked, with debug_info, not stripped
```

uninsane commented on pkgs/by-name/ha/hare/package.nix:
```diff
@@ -37,7 +71,14 @@ stdenv.mkDerivation (finalAttrs: {
-  ];
+  ] ++ lib.lists.optional enableCrossCompilation
+    # Set paths for cross-compilation toolchains
+    (substituteAll {
+      src = ./002-cross-compilation.patch;
+      inherit (tools.x86_64) x86_64_ld x86_64_as x86_64_cc;
+      inherit (tools.aarch64) aarch64_ld aarch64_as aarch64_cc;
+      inherit (tools.riscv64) riscv64_ld riscv64_as riscv64_cc;
+    });
```
maybe it makes more sense to do this substitution in preConfigure, where the logic is currently:
```nix
preConfigure = ''
  ln -s config.example.mk config.mk
'';
```

feels odd to be patching a source file named `example`.

onemoresuza replied:
Here are the `strings` output after such change on a `Hare` binary built for `x86_64`:

```console
[user@hostname:~/nixpkgs]$ strings result/bin/.hare-wrapped | grep -P '/nix/store/.*/(aarch64|riscv64)'

/nix/store/<hash>/bin/aarch64-unknown-linux-gnu-as
/nix/store/<hash>/bin/aarch64-unknown-linux-gnu-cc
/nix/store/<hash>/bin/aarch64-unknown-linux-gnu-ld
/nix/store/<hash>/bin/riscv64-unknown-linux-gnu-as
/nix/store/<hash>/bin/riscv64-unknown-linux-gnu-cc
/nix/store/<hash>/bin/riscv64-unknown-linux-gnu-ld


```

pbsds commented on pkgs/by-name/ha/hare/package.nix:
```diff
@@ -80,6 +133,14 @@ stdenv.mkDerivation (finalAttrs: {
   setupHook = ./setup-hook.sh;
 
+  passthru = {
+    tests = {
```
```suggestion
    tests = lib.optionalAttrs enableCrossCompilation {
```

onemoresuza replied:
Done

pbsds reviewed (commented):
Result of `nixpkgs-review pr 274565` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>1 package marked as broken and skipped:</summary>
  <ul>
    <li>himitsu-firefox</li>
  </ul>
</details>
<details>
  <summary>11 packages built:</summary>
  <ul>
    <li>hare</li>
    <li>hare.man</li>
    <li>hareThirdParty.hare-compress</li>
    <li>hareThirdParty.hare-ev</li>
    <li>hareThirdParty.hare-json</li>
    <li>hareThirdParty.hare-toml</li>
    <li>haredo</li>
    <li>haredo.man</li>
    <li>himitsu</li>
    <li>treecat</li>
    <li>treecat.man</li>
  </ul>
</details>

onemoresuza commented:
Result of `nixpkgs-review pr 274565` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>1 package marked as broken and skipped:</summary>
  <ul>
    <li>himitsu-firefox</li>
  </ul>
</details>
<details>
  <summary>14 packages built:</summary>
  <ul>
    <li>bonsai</li>
    <li>hare</li>
    <li>hare.man</li>
    <li>hareThirdParty.hare-compress</li>
    <li>hareThirdParty.hare-ev</li>
    <li>hareThirdParty.hare-json</li>
    <li>hareThirdParty.hare-png</li>
    <li>hareThirdParty.hare-ssh</li>
    <li>hareThirdParty.hare-toml</li>
    <li>haredo</li>
    <li>haredo.man</li>
    <li>himitsu</li>
    <li>treecat</li>
    <li>treecat.man</li>
  </ul>
</details>
